File: frontend/src/lib/disposition.ts
import { api } from './api'

export interface DispositionRequest {
  id: number
  documentId: number
  documentNumber: string | null
  documentTitle: string | null
  requestedAction: number          // 0 Destroy, 1 Renew
  requestedActionLabel: string
  reason: string
  requestedByUserId: number
  requestedByName: string | null
  requestedAtUtc: string
  status: number                   // 0 PendingVerification,1 PendingFinalApproval,2 Approved,3 Rejected,4 Completed
  statusLabel: string
  verifiedByUserId: number | null
  verifiedByName: string | null
  verifiedAtUtc: string | null
  verificationNotes: string | null
  finalApprovedByUserId: number | null
  finalApprovedByName: string | null
  finalApprovedAtUtc: string | null
  finalApprovalNotes: string | null
  rejectedByUserId: number | null
  rejectedByName: string | null
  rejectedAtUtc: string | null
  rejectionReason: string | null
  newExpiryDate: string | null
  method: number
  customMethod: string | null
  certificateId: number | null
  certificateNumber: string | null
  expiryDate: string | null
  boxCode: string | null
}

export interface DispositionCertificate {
  requestId: number
  certificateNumber: string
  documentIds: number[]
  documentNumbers: string[]
  destructionMethod: string
  verifiedByName: string | null
  finalApprovedByName: string | null
  generatedAtUtc: string
}

export interface Paged<T> { items: T[]; page: number; pageSize: number; totalCount: number; totalPages: number }

export const DISPOSITION_ACTION = { Destroy: 0, Renew: 1 } as const
export const DISPOSITION_STATUS = {
  PendingVerification: 0, PendingFinalApproval: 1, Approved: 2, Rejected: 3, Completed: 4,
} as const

/** Admin-named destruction methods (mirrors the backend DestructionMethod enum order). */
export const DESTRUCTION_METHODS: { value: number; label: string }[] = [
  { value: 0, label: 'محو تشفيري' },
  { value: 1, label: 'كتابة فوقية آمنة' },
  { value: 3, label: 'تقطيع' },
  { value: 4, label: 'حرق' },
  { value: 5, label: 'تذويب' },
  { value: 6, label: 'إزالة مغناطيسية' },
  { value: 7, label: 'أخرى' },
]

export const disposition = {
  list: (stage?: 'Verification' | 'FinalApproval', page = 1, pageSize = 50) =>
    api.get<Paged<DispositionRequest>>('/disposition-requests', { params: { stage, page, pageSize } }).then((r) => r.data),
  get: (id: number) => api.get<DispositionRequest>(`/disposition-requests/${id}`).then((r) => r.data),
  create: (b: { documentId: number; requestedAction: number; reason: string; method?: number; customMethod?: string | null }) =>
    api.post<DispositionRequest>('/disposition-requests', b).then((r) => r.data),
  verify: (id: number, decision: 'Verify' | 'Reject', notes: string) =>
    api.post<DispositionRequest>(`/disposition-requests/${id}/verify`, { decision, notes }).then((r) => r.data),
  finalApprove: (id: number, decision: 'Approve' | 'Reject', notes: string, newExpiryDate?: string | null) =>
    api.post<DispositionRequest>(`/disposition-requests/${id}/final-approve`, { decision, notes, newExpiryDate }).then((r) => r.data),
  reject: (id: number, reason: string) =>
    api.post<DispositionRequest>(`/disposition-requests/${id}/reject`, { reason }).then((r) => r.data),
  certificate: (id: number) =>
    api.get<DispositionCertificate>(`/disposition-requests/${id}/certificate`).then((r) => r.data),
}

File: backend/src/Archiving.Domain/Entities/Disposition.cs
using Archiving.Domain.Common;
using Archiving.Domain.Enums;

namespace Archiving.Domain.Entities;

/// <summary>Per-document retention tracking. Created when a policy is assigned (or by the background job),
/// it drives expiry detection and the two-step disposition workflow.</summary>
public class DocumentRetention : BaseEntity
{
    public long DocumentId { get; set; }
    public long? RetentionPolicyId { get; set; }
    /// <summary>The resolved date the retention clock started (per the policy's trigger type).</summary>
    public DateOnly TriggerDate { get; set; }
    public DateOnly ExpiryDate { get; set; }
    public DocumentRetentionStatus Status { get; set; } = DocumentRetentionStatus.Active;
    /// <summary>Preserved across renewals for audit (the first/original expiry).</summary>
    public DateOnly? OriginalExpiryDate { get; set; }

    public Document Document { get; set; } = null!;
    public RetentionPolicy? RetentionPolicy { get; set; }
}

/// <summary>A two-step disposition request (Destroy or Renew) raised when a document reaches end-of-retention.
/// Verification (Records Officer) → Final Approval (Legal/Department Head). The row is retained permanently.</summary>
public class DispositionRequest : BaseEntity
{
    public long DocumentId { get; set; }
    public DateTime RequestedAtUtc { get; set; } = DateTime.UtcNow;
    public DispositionAction RequestedAction { get; set; } = DispositionAction.Destroy;
    public string Reason { get; set; } = string.Empty;
    public long RequestedByUserId { get; set; }
    public DispositionStatus Status { get; set; } = DispositionStatus.PendingVerification;

    // Step 1 — verification (Records Officer)
    public long? VerifiedByUserId { get; set; }
    public DateTime? VerifiedAtUtc { get; set; }
    public string? VerificationNotes { get; set; }

    // Step 2 — final approval (Legal / Department Head)
    public long? FinalApprovedByUserId { get; set; }
    public DateTime? FinalApprovedAtUtc { get; set; }
    public string? FinalApprovalNotes { get; set; }

    // Rejection (at either step)
    public long? RejectedByUserId { get; set; }
    public DateTime? RejectedAtUtc { get; set; }
    public string? RejectionReason { get; set; }

    // Renew outcome
    public DateOnly? NewExpiryDate { get; set; }

    // Destroy outcome
    public DestructionMethod Method { get; set; } = DestructionMethod.CryptoShred;
    public string? CustomMethod { get; set; }
    public long? CertificateId { get; set; }

    public Document Document { get; set; } = null!;
}

/// <summary>Certificate of Destruction, signed by both the verifier and the final approver. Batch-capable
/// (one certificate may cover several documents/requests destroyed together).</summary>
public class DispositionCertificate : BaseEntity
{
    public long DispositionRequestId { get; set; }
    public string CertificateNumber { get; set; } = string.Empty;
    /// <summary>Comma-separated document ids covered by this certificate (batch-capable).</summary>
    public string DocumentIds { get; set; } = string.Empty;
    public string DestructionMethod { get; set; } = string.Empty;
    public long? VerifiedByUserId { get; set; }
    public long? FinalApprovedByUserId { get; set; }
    public DateTime GeneratedAtUtc { get; set; } = DateTime.UtcNow;
    public string? PdfStorageKey { get; set; }
}
